overlord/hookstate/hookmgr.go

```diff
@@ -253,6 +272,29 @@ func (m *HookManager) doRunHook(task *state.Task, tomb *tomb.Tomb) error {
 		return err
 	}
 
+	return m.runHook(task, tomb, snapst, hooksup)
+}
+
+// undoRunHook runs the undo-hook that was requested.
+//
+// Note that this method is synchronous, as the task is already running in a
+// goroutine.
+func (m *HookManager) undoRunHook(task *state.Task, tomb *tomb.Tomb) error {
+	task.State().Lock()
+	hooksup, snapst, err := undoHookSetup(task)
+	task.State().Unlock()
+	if err != nil {
+		if err == state.ErrNoState {
```

Comment:
We're even already checking this externally (which is good). This would never happen with the current shape of the function.